test/e2e/discovery/service_discovery.go

```diff
@@ -317,6 +387,62 @@ func verifyServiceIpWithDig(f *framework.Framework, srcCluster, targetCluster fr
 	})
 }
 
+func verifyRoundRobinWithDig(f *framework.Framework, srcCluster framework.ClusterIndex, serviceName string, serviceIPList []string,
+	targetPod *corev1.PodList, domains []string, clusterName string, shouldContain bool) {
+	cmd := []string{"dig", "+short"}
+	var clusterDNSName string
+	if clusterName != "" {
+		clusterDNSName = clusterName + "."
+	}
+
+	for i := range domains {
+		cmd = append(cmd, clusterDNSName+serviceName+"."+f.Namespace+".svc."+domains[i])
+	}
+
+	op := "is"
+	if !shouldContain {
+		op += " not"
+	}
+
+	serviceIPMap := make(map[string]int)
+
+	By(fmt.Sprintf("Executing %q to verify IPs %q for service %q %q discoverable", strings.Join(cmd, " "), serviceIPList, serviceName, op))
+
+	for count := 0; count < 10; count++ {
+		framework.AwaitUntil("verify if service IP is discoverable", func() (interface{}, error) {
+			stdout, _, err := f.ExecWithOptions(framework.ExecOptions{
+				Command:       cmd,
+				Namespace:     f.Namespace,
+				PodName:       targetPod.Items[0].Name,
+				ContainerName: targetPod.Items[0].Spec.Containers[0].Name,
+				CaptureStdout: true,
+				CaptureStderr: true,
+			}, srcCluster)
+			if err != nil {
+				return nil, err
+			}
+
+			return stdout, nil
+		}, func(result interface{}) (bool, string, error) {
+			var doesContain bool
+			for _, serviceIp := range serviceIPList {
+				doesContain = strings.Contains(result.(string), serviceIp)
+				if doesContain {
+					serviceIPMap[serviceIp]++
+				}
+			}
+			By(fmt.Sprintf("Validating that dig result %s %q", op, result))
+			return true, "", nil
+		})
+	}
+
+	By(fmt.Sprintf("Validating that difference between the service IP %q which was returned %d times and  the service"+
+		" IP %q which was returned %d times is within the threshold", serviceIPList[0], serviceIPMap[serviceIPList[0]],
+		serviceIPList[1], serviceIPMap[serviceIPList[1]]))
+
+	Expect(serviceIPMap[serviceIPList[0]]-serviceIPMap[serviceIPList[1]]%10 < 5).Should(BeTrue())
```

Comment:
I'm not clear on the algorithm/calculation here. We expect the IPs to be interleaved, ie IP1, IP2, IP1, IP2 etc.  If all 10 requests returned IP1, then the calculation `(10 - 0) % 10 = 0` would be < 5 which would pass, as would `countIP1 = 7` and `countIP2 = 3`.  Even if both IPs were returned 5 times, this wouldn't verify they were properly interleaved. I suggest:

```
    retIPs := make([]string, 0)
    seenIPs := stringset.New()
    ipCount := len(serviceIPs)
    totalReqCount := ipCount * 3
    for count := 0; count < totalReqCount; count++ {
        framework.AwaitUntil("verify round-robin service IP retrieval", func() (interface{}, error) {
	    ...
        }, func(result interface{}) (bool, string, error) {
	    output := result.(string)
	    isValidIP := false
	    for _, serviceIP := range serviceIPs {
	        if strings.Contains(output, serviceIP) {
		    if count < ipCount && !seenIPs.Add(serviceIP) {
		        return false, "", fmt.Errorf("dig returned IP %s out of round-robin order - prior IPs %v",
			    serviceIP, seenIPs.Elements())
		    }

		    retIPs = append(retIPs, serviceIP)
		    isValidIP = true
		    break
		}
             }

	    if !isValidIP {
	        return false, "", fmt.Errorf("dig result %q does not contain any expected IP %v", output, serviceIPs)
	    }   

	    return true, "", nil
	})
    }

    rrIPs := retIPs[0:ipCount]
    for i := ipCount; i < totalReqCount; i += ipCount {
        Expect(retIPs[i:i+ipCount]).To(Equal(rrIPs), "Service IPs returned out of round-robin order: %v", retIPs)
    }

```

Reply:
It may not return in exact round-robin order as there are two lighthouse core-dns instance running and any one of them could return the service IP for a  DNS request. Each of them maintains it own data structures. That is the reason for comparing it with a threshold.

Reply:
OK. In talking with Vishal:

- we should take the absolute value of the difference, ie `int(math.Abs(serviceIPMap[serviceIPList[0]]-serviceIPMap[serviceIPList[1]])( % 10`
- to avoid `(10 - 0) % 10 = 0` from passing, mod by 11 instead
- use threshold 3 instead of 5 as the most times the same IP should be returned out of 10 requests is 6 - so 7 out of 10 should fail (`(7-3) % 10 = 4`).

Reply:
we require either "%" or Abs. I just assumed % would always return a positive value. Now changed to use abs.